### DISCUSSION TITLE
Animations

### DISCUSSION THREAD

**Discussion by TopperDEL** [StereoKit/StereoKit#120] [Q&A] [answered]
Do you know if there is any animation library, that could be used together with StereoKit? I'm searching for little reuse animations for "popping in" and "removing" of elements. For 2D frameworks there exist sometimes similar animations if an element gets added to a list and such. I'm wondering if there exists some reusable  library or similar to bring some live the the app.

Any ideas?

**Comment by maluoi** [comment, marked as answer]
Like a Tweening library? I know of a few that'd work for Unity (DOTween and iTween come to mind), but I'm not familiar with platform agnostic ones. I've written a few in the past for my own stuff, they're usually pretty simple and fun to make!

I found [this one](https://gist.github.com/nickgravelyn/4953988) that's pretty small, and might be easy enough to port to work with StereoKit? If this isn't what you're thinking of though, let me know!

Writing a Tweening library is a fun idea, and I definitely would like to build one for StereoKit when I get the time :)

**Reply by TopperDEL** [reply]
Tweening was the word I was searching for! Yeah, I could have remembered this from the good old Flash-Designer. [This](https://github.com/eropple/UnglidePCL) library seems to be a good looking one, I guess. All I would need for this is an "elapsed-time"-value. I will give it a shot.

Another question in this field: can I set transparency on all rendered objects? My goal is to animate the entering of a menu consisting of some buttons. They should near the user a bit and get opaque in the same time. "nearing" the user can be achieved by adjusting the position - for the fade-in I would need to adjust an "opaqueness/transparency"-value. Is that possible?

**Reply by TopperDEL** [reply]
Thank you so much for your helpful an in-depth-answers!

**Reply by maluoi** [reply]
For time, check out `Time.Elapsedf` and `Time.Totalf`!

Regarding transparency, the Material system is capable of doing this! It might be tricky to get UI elements to do it, since they share a single material, and manage their own colors. If you set `Material.Transparency` to `Transparency.Blend`, then the material will use the alpha color for transparency! The final alpha value varies a little from shader to shader, but is generally something like this: `TextureAlpha * VertexColorAlpha * MaterialColorTintAlpha * RenderInstanceAlpha`, where the first two are from art assets, MaterialColorTintAlpha can be assigned via `material[MatParamName.ColorTint] = Color.White`, and RenderInstanceAlpha is assigned through the last parameter of the render call `Renderer.Add(model, Matrix.Identity, Color.White)`.

However, just remember that transparency opens a big box of worms! Depth based Late Stage Reprojection requires accurate depth values, and transparent materials tend to work best when you turn off depth-write. You also have to start worrying about draw-order issues, which is made even trickier since StereoKit doesn't currently z-sort transparent objects for the painter's algorithm! If you can achieve your effects without transparency, that's always the best option. Scaling, fading to black, stuff like that will make your life much easier!
